xdp-dump/xdpdump.c

```diff
@@ -416,94 +417,30 @@ static bool capture_on_legacy_interface(struct dumpopt *cfg)
 }
 
 /*****************************************************************************
- * guess_full_main_function()
+ * find_target_program()
  *****************************************************************************/
-static char *guess_full_main_function(struct bpf_prog_info *info,
-				      char *function_override)
+static struct xdp_program *find_target_program(struct xdp_multiprog *mp,
+					       char *function_override)
 {
-	int         nr_types;
-	int         matches = 0;
-	char       *func_name_ptr = NULL;
-	struct btf *btf = NULL;
-
-	/*
-	 * If the function name is longer than 15 characters the bpf_prog_info
-	 * structure will truncate it to 15 characters. Here we try to see if
-	 * we need to check BTF information to figure out the full function
-	 * name.
-	 */
-	if ((strlen(info->name) < (BPF_OBJ_NAME_LEN - 1)
-	     && !function_override) ||
-	    btf__get_from_id(info->btf_id, &btf))
-		return strdup(function_override ? function_override :
-			      info->name);
-
-
-	nr_types = btf__get_nr_types(btf);
-	for (int i = 1; i <= nr_types; i++) {
-		const        char     *name;
-		const struct btf_type *t = btf__type_by_id(btf, i);
-
-		if (!btf_is_func(t))
-			continue;
+	struct xdp_program *prog = NULL;
 
-		name = btf__name_by_offset(btf, t->name_off);
-		if (name) {
-			if (function_override) {
-				if (!strcmp(function_override, name)) {
-					func_name_ptr = (char *)name;
-					matches++;
-					break;
-				}
-			} else if (!strncmp(info->name, name,
-					    BPF_OBJ_NAME_LEN - 1)) {
-				func_name_ptr = (char *)name;
-				matches++;
-			}
-		}
-	}
+	if (!function_override)
+		return xdp_multiprog__dispatcher(mp);
 
-	if (matches == 0) {
-		pr_warn("ERROR: Can't find function, %s, in BTF information!\n",
-			function_override ? function_override : info->name);
-		func_name_ptr = NULL;
-	} else if (matches > 1) {
-		pr_warn("ERROR: Can't identify the full XDP main function!\n"
-			"The following is a list of candidates:\n");
-
-		for (int i = 1; i <= nr_types; i++) {
-			const        char     *name;
-			const struct btf_type *t = btf__type_by_id(btf, i);
-
-			if (!btf_is_func(t))
-				continue;
-
-			name = btf__name_by_offset(btf, t->name_off);
-			if (name &&
-			    !strncmp(info->name, name, BPF_OBJ_NAME_LEN - 1))
-				pr_warn("  %s\n", name);
-		}
+	while ((prog = xdp_multiprog__next_prog(prog, mp)))
+		if (!strcmp(function_override, xdp_program__name(prog)))
+			return prog;
 
-		pr_warn("Please use the -p option to pick the correct one.\n");
-		func_name_ptr = NULL;
-	} else {
-		func_name_ptr = strdup(func_name_ptr);
-	}
-
-	btf__free(btf);
-	return func_name_ptr;
+	return NULL;
```

Comment:
Will schedule a full review of these changes for early next week, but in this case we should still have something like thise: 

```
pr_warn("ERROR: Can't find function, %s, in BTF information!\n"
```

Reply:
What? Telling the user what went wrong instead of just silently erroring out and leaving them guessing? Luxury, I say! ;)